File: Apps/Chat/android/chatkit/src/main/kotlin/com/dweekly/cyrinx/chat/ChatEvent.kt
package com.dweekly.cyrinx.chat

/**
 * Every event carries [eventSeq], monotonic from 0, per client instance (client
 * A's and client B's `eventSeq` sequences are both independently zero-based and
 * never compared to each other). Eight payload kinds. ../../../CONTRACT.md
 * section 1.7.
 *
 * No event is emitted for a client's implicit initial state (no peers,
 * `disconnected(reason: null)`) -- events represent transitions, not the zero
 * state.
 *
 * DECISION (not pinned by the brief, per ../../../CONTRACT.md section 1.7): pinned
 * here as a shared abstract property realized per-subclass (idiomatic Kotlin,
 * since sealed-class hierarchies commonly hoist a shared field to the base class
 * rather than wrapping) rather than a wrapper struct pairing `eventSeq` with a
 * nested payload enum (the Swift shape). The semantic content -- eventSeq plus
 * exactly these eight payload shapes with exactly these fields -- is identical on
 * both platforms; only which declaration carries `eventSeq` differs.
 */
sealed class ChatEvent {
    abstract val eventSeq: Long

    data class PeerFound(override val eventSeq: Long, val peer: ChatPeer) : ChatEvent()

    data class PeerUpdated(override val eventSeq: Long, val peer: ChatPeer) : ChatEvent()

    data class PeerLost(
        override val eventSeq: Long,
        val peerIdHex: String,
        val reason: String,
    ) : ChatEvent()

    data class ConnectionChanged(
        override val eventSeq: Long,
        val state: ChatConnectionState,
    ) : ChatEvent()

    data class LinkBudgetChanged(
        override val eventSeq: Long,
        val budget: ChatLinkBudget,
    ) : ChatEvent()

    data class MessageReceived(override val eventSeq: Long, val message: ChatMessage) : ChatEvent()

    data class MessageStatusChanged(
        override val eventSeq: Long,
        val messageIdHex: String,
        val status: ChatMessageDisplayStatus,
    ) : ChatEvent()

    data class ClientFailed(override val eventSeq: Long, val reason: String) : ChatEvent()
}

File: Apps/Chat/CyrinxChatKit/Sources/CyrinxChatKit/ChatEvent.swift
import Foundation

/// A single transport-emitted event, wrapping a monotonic `eventSeq` around
/// one of eight payload kinds. Apps/Chat/CONTRACT.md §1.7.
///
/// `eventSeq` is monotonic from 0, **per client instance** -- client A's and
/// client B's `eventSeq` sequences are both independently zero-based and
/// never compared to each other. No event is emitted for a client's
/// implicit initial state (no peers, `disconnected(reason: nil)`) -- events
/// represent *transitions*, not the zero state.
///
/// CONTRACT.md's own "DECISION (not pinned by the brief)" note: the brief
/// under-specifies whether `eventSeq` is a field on a wrapper type or
/// duplicated onto every case. CONTRACT.md pins Swift as a wrapper struct
/// (this type); Kotlin instead hoists `eventSeq` onto a sealed-class base
/// property. The semantic content is identical on both platforms.
public struct ChatEvent: Equatable, Sendable {
    public let eventSeq: UInt64
    public let kind: Kind

    public init(eventSeq: UInt64, kind: Kind) {
        self.eventSeq = eventSeq
        self.kind = kind
    }

    public enum Kind: Equatable, Sendable {
        case peerFound(ChatPeer)
        case peerUpdated(ChatPeer)
        case peerLost(peerIdHex: String, reason: String)
        case connectionChanged(ChatConnectionState)
        case linkBudgetChanged(ChatLinkBudget)
        case messageReceived(ChatMessage)
        case messageStatusChanged(messageIdHex: String, status: ChatMessageDisplayStatus)
        case clientFailed(reason: String)
    }
}
